### ISSUE TITLE
Selecting GPUs and compute backends

### ISSUE THREAD

**Issue by farooqkz** [Jeffser/Alpaca#428]
**Is your feature request related to a problem? Please describe.**

These days a good number of computers come with multiple GPUs, and each GPU usually has got multiple APIs to offload computation to. For instance, Most GPUs support both OpenCL and Vulkan for compute. Intel Arc GPUs also have got XMX cores specifically created for matrix operations. Furthermore, many laptops and workstations come with more than a single GPU.

**Describe the solution you'd like**

It would be nice to have a selection mechanism for selecting GPU or CPU. And if it's a GPU, which GPU if there are multiple GPUs. There should be a global setting and users should be able to override the global one per model. Sensible defaults should be there. Such as always trying to use a GPU if one is available.

**Describe alternatives you've considered**

Not much of an alternative comes to mind.

**Additional context**
Additional to GPUs, some PCs, even cheap SBCs come with an NPU or TPU. This should be kept in mind.

**Comment by Jeffser** [comment]
Hi, the backend for Alpaca is Ollama, as far as I know Ollama is only compatible with ROCm and Cuda (no Vulkan / OpenCL)

Alpaca does provide some options in the preferences dialog to modify which GPU is being used, for example CUDA_VISIBLE_DEVICES and HIP_VISIBLE_DEVICES

**Comment by Jeffser** [comment]
I just noticed Ollama doesn't use HIP_VISIBLE_DEVICES anymore, I'm going to fix that

**Comment by Jeffser** [comment]
https://github.com/Jeffser/Alpaca/commit/4d50b7d008586be95a44b3d4335e0b74f5ce36dc

Alright that fixed,

I could make a GPU / CPU selector in the preferences dialog but it could take a while to get it working correctly

For now, using those overrides is the best way of configuring Ollama

**Comment by ossenthi** [comment]
I haven't use the program for while, but I started using it again about a week ago and noticed it doesn't use my Nvidia RTX mobile Max-Q Gpu anymore. I Have tried everything guide told me to. I have cuda installed and works for other ollama-cuda setup. we need better way to diagnose problems when it comes to fails to utilize gpu.

**Comment by farooqkz** [comment]
> Hi, the backend for Alpaca is Ollama, as far as I know Ollama is only compatible with ROCm and Cuda (no Vulkan / OpenCL)
> 
> Alpaca does provide some options in the preferences dialog to modify which GPU is being used, for example CUDA_VISIBLE_DEVICES and HIP_VISIBLE_DEVICES

I think Vulkan for Ollama is in WIP status. I saw some PR somewhere.

**Comment by stereomato** [comment]
Ollama also supports Intel thru oneAPI.

**Comment by farooqkz** [comment]
Vulkan is vendor neural. There are tons of GPUs which support Vulkan. There are ARM GPUs for instance which usually work only with OpenCL and Vulkan.

**Comment by Jeffser** [comment]
There's nothing I can do from Alpaca's side to support vulkan that's for Ollama to work on and they don't seem interested in that.

**Comment by padoruuuu** [comment]
could you make it possible to select the integrated runner then? https://github.com/pufferffish/ollama-vulkan id love to be able to change in the gui to ollama-vulkan

**Comment by stereomato** [comment]
@padoruuuu how do you even build that?

**Comment by Jeffser** [comment]
> could you make it possible to select the integrated runner then? https://github.com/pufferffish/ollama-vulkan id love to be able to change in the gui to ollama-vulkan

well if you manage to get that running you could connect to it using the option to connect to remote instances in preferences

**Comment by padoruuuu** [comment]
> > could you make it possible to select the integrated runner then? https://github.com/pufferffish/ollama-vulkan id love to be able to change in the gui to ollama-vulkan
> 
> well if you manage to get that running you could connect to it using the option to connect to remote instances in preferences


honestly i never realized that remote connection meant connecting to locally running models this is the exact functionality i was asking for.

also i just used ai to make a quick and lazy pkgbuild for ollama-vulkan and its working fine. thank you!

**Comment by stereomato** [comment]
yeah, im using a docker container (ollama-intel-arc) to run ollama with oneapi for gpu accel myself

Sent from Proton Mail Android

-------- Original Message --------

**Comment by farooqkz** [comment]
> There's nothing I can do from Alpaca's side to support vulkan that's for Ollama to work on and they don't seem interested in that.

What about supporting backends other than ollama?

**Comment by Jeffser** [comment]
You can use any OpenAI compatible instance by going to `instance manager` and selecting the instance you want (or `OpenAI compatible instance` in case it isn't listed), I believe llama.cpp should work with it.

**Comment by bellegarde-c** [comment]
https://github.com/ollama/ollama/pull/11835
